### ISSUE TITLE
NixOS support

### ISSUE THREAD

**Issue by 0x4A6F** [pr1ntf/iohyve#193]
I am aware of two ways to boot NixOS in iohyve.
1. Using generationsDir and custom _grub.cfg_
2. Using EFI boot.

The second one isn't yet available on FreeBSD-RELEASE and I've successfully booted and installed NixOS with the first method following these instructions. 

Download iso from https://nixos.org/nixos/download.html
and find the arguments for grub. For example mount the iso and examine the APPEND line in _isolinux/isolinux.cfg_
.
This is an example for my downloaded [iso](https://nixos.org/releases/nixos/16.03-small/nixos-16.03.802.01f5711/nixos-minimal-16.03.802.01f5711-x86_64-linux.iso) (adjust your init accordingly):

```
linux (cd0)/boot/bzImage init=/nix/store/1syxa5b95j08nxbnshjkf3d9q274dzgd-nixos-system-nixos-16.03.802.01f5711/init root=LABEL=NIXOS_ISO boot.shell_on_fail loglevel=7
initrd (cd0)/boot/initrd
boot
```

Fetch iso, create and install nixos-guest.

```
# iohyve fetch https://nixos.org/releases/nixos/latest-iso-minimal-x86_64-linux
# iohyve create nixos 16G && iohyve set nixos loader=grub-bhyve && iohyve set nixos loader=custom
# iohyve install nixos latest-iso-minimal-x86_64-linux
# iohyve console
```

Use the grub commands extracted from isolinux.cfg as shown above and be aware that latest-iso is a moving target.

Enable Network:

```
# echo "nameserver ${address-nameserver}" >> /etc/resolv.conf
# ifconfig enp0s3 inet ${address}/${netmask}
# route add default gw {address-gateway} enp0s3
```

Format disk (4G swap and remaining space for root):

```
# fdisk /dev/sda
# mkswap -L swap /dev/sda1
# swapon /dev/sda1
# mkfs.ext4 -L nixos /dev/sda2
# mount /dev/disk/by-label/nixos /mnt
# nixos-generate-config --root /mnt
# nano /mnt/etc/nixos/configuration.nix
```

Configure NixOS, uncomment _boot.loader.grub.devices_ and add:

```
  boot.loader.generationsDir.copyKernels = true;
  boot.loader.generationsDir.enable = true;
```

Install NixOS:

```
# nixos-install
# reboot
```

Populate _/iohyve/nixos/grub.cfg_:

```
set root=(hd0,msdos2)
linux /boot/nixos-kernel systemConfig=/boot/default/system init=/boot/nixos-init loglevel=4
initrd /boot/nixos-initrd
boot
```

and _/iohyve/nixos/device.map_:

```
(hd0) /dev/zvol/zroot/iohyve/nixos/disk0
```

But one problem remains. `iohyve start nixos` will not boot the supplied _grub.cfg_. 
Manually entering lines from _grub.cfg_ works, if os is set to default, but not if set to custom.
`iohyve set nixos os=default` 

Tested with iohyve v0.7.3 2016/01/08 Bear in a Datacenter Edition

**Comment by pr1ntf** [comment]
> Tested with iohyve v0.7.3 2016/01/08 Bear in a Datacenter Edition

The custom `grub.cfg` stuff did not become available until version `0.7.5 Tennessee Cherry Moonshine Edition` which is currently in [ports](https://www.freshports.org/sysutils/iohyve/). Version `0.7.6` should be available in ports as soon as it's [accepted into the ports tree](https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=210546).

You can also install the github version if you'd like. 

```
git clone https://github.com/pr1ntf/iohyve.git

cd iohyve

sudo make install clean
```

**Comment by pr1ntf** [comment]
If you have any further issues regarding NixOS please feel free to reopen this issue or open another.
